plugins/inputs/procstat/connections_linux.go

```diff
@@ -0,0 +1,400 @@
+* about procs networking.
+* The main idea is to add outgoing and listeners for each process.
+* Procs in others network namespaces other than the default are taked into account just for outgoing
+* connections, as listeners are not accessible, so ignored.
+* The main flow of this functions is:
+*   - read all /proc/ pid directories to get inodes and network namespaces
+*   - for each network NS, call Netlink to get all connections
+*   - map those connections to procs (using inode)
+*   - return also the list of listeners ports (only in the default network ns)
+ */
+
+package procstat
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"runtime"
+	"strconv"
+	"strings"
+
+	"github.com/elastic/gosigar/sys/linux"
+	"github.com/vishvananda/netns"
+)
+
+// inode numeric representation of inodes
+type inode uint32
+
+// netNSProcInfo stores proc info grouped by network namespace
+type netNSProcInfo struct {
+	data map[string]netNSProcs
+}
+
+// netNSProcs store for each network namespace, a link to its path and the list
+// of procs under it
+type netNSProcs struct {
+	id string
+	// inodeToProc map each inode to a list of procs (pid + ppid)
+	inodeToProc map[inode][]inodeInfo
+}
+
+// connInfo represents a single proc's connection and the parent pid (for
+// practical purposes ).
+type connInfo struct {
+	netNs   string
+	state   linux.TCPState
+	srcIP   net.IP
+	srcPort uint32
+	dstIP   net.IP
+	dstPort uint32
+}
+
+// IsAListenPort returns true if the port param is associated with a listener
+// found in the host connections
+func (n *networkInfo) IsAListenPort(port uint32) bool {
+	_, ok := n.listenPorts[port]
+	return ok
+}
+
+// Fetch fetches network info: TCP connections and hosts' IPs.
+// Parameter getConnections is the function that will be used to obtain TCP
+// connections.
+// Parameter getLocalIPs is the function that will be used to get IPs.
+// It is passed as a parameter to facilitate testing.
+func (n *networkInfo) Fetch() error {
+	var err error
+	n.tcp, n.listenPorts, err = n.getTCPProcInfo()
+	if err != nil {
+		return fmt.Errorf("TCP info: %w", err)
+	}
+
+	// Get IPs, to be able to resolve procs listening in 0.0.0.0 or ::
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		return fmt.Errorf("network interfaces: %w", err)
+	}
+
+	n.publicIPs, n.privateIPs, err = getLocalIPs(ifaces)
+	if err != nil {
+		return fmt.Errorf("local IPs: %w", err)
+	}
+
+	return nil
+}
+
+// GetConnectionsByPid return connection info for a particular PID
+func (n *networkInfo) GetConnectionsByPid(pid uint32) ([]connInfo, error) {
+	conn, ok := n.tcp[pid]
+	if !ok {
+		return conn, errPIDNotFound
+	}
+	return conn, nil
+}
+
+// GetPublicIPs return the list of public IPs (used to connect to others hosts)
+func (n *networkInfo) GetPublicIPs() []net.IP {
+	return n.publicIPs
+}
+
+// GetPrivateIPs return the list of private IPs (loopback devices, virtual,
+// point2point)
+func (n *networkInfo) GetPrivateIPs() []net.IP {
+	return n.privateIPs
+}
```

Comment:
I would really rather access the information directly from the struct as it is internal anyway...